### DISCUSSION TITLE
Integrating with an existing dapp/web3provider

### DISCUSSION THREAD

**Discussion by sirpy** [celo-org/react-celo#328] [Q&A]
We have an existing dapp that uses other frameworks to connect a wallet etc.
And we'd like to add the ubeswap component inside our dapp.
How can I pass/set the web3 provider for CeloProvider to the existing web3provider/jsonrpcprovider that the user is already connected to?
In other words, I don't to trigger the "connect wallet modal" again for the user via the celokit, because the user is already connected via our own framework.

**Comment by aaronmgdr** [comment]
yeah my apologies but there isnt a way to do this. It wouldnt be something we would add to react-celo (even if we did ubeswap uses its own fork)  

It would be best in my opinion if the ubeswap component just asked for a  signer / provider to be passed in and didnt include any wallet connection so that dapps like yours didnt need to load 2 of them

**Comment by aaronmgdr** [comment]
If i was in your position id look at forking the Swap Component repo

**Reply by sirpy** [reply]
Thanks.
1. I think this is crucial for any future integrations between dapps, the ability to easily create a provider context, we've built ours this way
2. Ubeswap component has been updated to use the main version and not a fork
3. Forking ubeswap would seem to me as very time consuming as it would require changing lots of celokit code dependency
4. Isn't there a way to simply update the context state with a provider? That doesn't seem unreasonable. Just like after a wallet connects and it sets the provider. I've tried to look at the kit code but it was complex to follow the logic. If someone could point me to the right place I could probably create a PR. Are there specific complexities you see there?

**Reply by aaronmgdr** [reply]
To me this is a discussion that would be better to have on the ubeswap component repo. 

my view is that ContextProviders like the CeloProvider from the react-celo package work best when they are able to wrapp the entire app and thus share data anywhere. with the `<Swap/>` component the Context is inside of it. 

If instead the Swap did something like 
```
<CeloProvider>
</Swap />
</CeloProvider
```
then other Components could be in there too 

```
<CeloProvider>
</Swap />
<YourDappComponent/>
</CeloProvider
```

I think Passing provider in as a prop to CeloProvider would be buggy at best.

**Reply by sirpy** [reply]
1. Yes but that doesnt matter, even if the `<celoprovider>` would be outside, we would still need to pass it our own web3provider instance to prevent double wallet connection popup.
by not allowing the passing a web3provider you are locking in dapps instead of making their components shareable in other dapps
our dapp has its own context provider based on useDapp
The following is a much more modular and open architecture.
```
<G$Provider>
  <UseDAPPProvider provider=somejsonrpcprovider>
      <OurDapp>
         <CeloProvider provider=somejsonrpcprovider>
           <Swap/>
        </CeloProvder>
        <SomeOtherProvider provider=somejsonrpcprovider>
            <SomeOtherDapp/>
       </SomeOtherProvider>
  </OurDAPP>
 </UseDAPPProvider>
</G$Provider> 
```
If someone wanted to use our components he would just have to a provider `<G$Provider provider=somejsonrpcprovider>`

2. Why would it be buggy? we are only talking about skipping the wallet connection part, the rest is purely using an ethers provider.
all functionality including switchchain should just work. instead of getting the provider from your wallet connectors you are getting an existing similar provider. Think of it as adding another wallet option that just doesnt trigger the modal and immediately returns a provider.

**Reply by sirpy** [reply]
maybe jsonrpcprovider is not the right term, there's some erc standard for the provider i forgot.

**Reply by aaronmgdr** [reply]
Im starting to see your point here. Having an external Provider as a prop bypass the wallet part. 

Let me see what our PM and team think and if so when we could slide it in.

**Reply by aaronmgdr** [reply]
I do worry your dapp bundle size will be absolutely massive if you do this too many times

**Reply by sirpy** [reply]
Awesome thanks. Again happy to make a PR if you can point me to the provider setting logic and api
Bundle size depends on many factors I don't see why this specifically would make a huge bundle size.
